test/cryptography/EIP712.ts

```diff
@@ -0,0 +1,49 @@
+import { hashData, signData } from '@solidstate/library';
+import { EIP712Mock, EIP712Mock__factory } from '@solidstate/typechain-types';
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+
+describe('EIP712', function () {
+  let instance: EIP712Mock;
+
+  beforeEach(async function () {
+    const [deployer] = await ethers.getSigners();
+    instance = await new EIP712Mock__factory(deployer).deploy();
+  });
+
+  describe('__internal', function () {
+    describe('#calculateDomainSeparator(bytes32,bytes32)', function () {
+      it('calculates EIP-712 domain separator', async function () {
+        const typeHash = ethers.utils.solidityKeccak256(
+          ['string'],
+          [
+            'EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)',
+          ],
+        );
+
+        const nameHash = ethers.utils.solidityKeccak256(['string'], ['name']);
+        const versionHash = ethers.utils.solidityKeccak256(['string'], ['1']);
+
+        const chainId = await ethers.provider.send('eth_chainId');
+
+        const domainSeparator = ethers.utils.solidityKeccak256(
+          ['bytes32', 'bytes32', 'bytes32', 'uint256', 'address'],
+          [
+            typeHash,
+            nameHash,
+            versionHash,
+            chainId,
+            ethers.utils.hexZeroPad(instance.address, 32),
```

Comment:
@0xCourtney I don't understand why the `hexZeroPad` is needed here.  Any idea?

Reply:
`solidityKeccak256` is packing values. 

This would work :
```ts
        import { defaultAbiCoder, keccak256 } from 'ethers/lib/utils';

        const domainSeparator = keccak256(defaultAbiCoder.encode(['bytes32', 'bytes32', 'bytes32', 'uint256', 'address'], [
            typeHash,
            nameHash,
            versionHash,
            chainId,
            instance.address,
          ]));
```

Reply:
In solidity, the domain is constructed without packed encoding (ie `abi.encode`), according to the `ethers` [docs](https://docs.ethers.org/v5/api/utils/hashing/#utils-solidityKeccak256) `solidityKeccak256` packs encoded values, as @Lumyo mentioned.

> Returns the [KECCAK256](https://en.wikipedia.org/wiki/SHA-3) of the non-standard encoded values packed according to their respective type in types.

We should probably use `keccak256` in this test to be consistent. 

![Screenshot 2023-05-24 at 11 16 24](https://github.com/solidstate-network/solidstate-solidity/assets/23440493/12a588f4-44c4-4c81-82c6-45df2d682149)